lib/puppet/type/file.rb

```diff
@@ -324,20 +324,22 @@ def self.title_patterns
     defaultto '%'
   end
 
-  # Autorequire the nearest ancestor directory found in the catalog.
-  autorequire(:file) do
-    req = []
-    path = Pathname.new(self[:path])
-    if !path.root?
-      # Start at our parent, to avoid autorequiring ourself
-      parents = path.parent.enum_for(:ascend)
-      if found = parents.find { |p| catalog.resource(:file, p.to_s) }
-        req << found.to_s
+  # Autorequire the nearest ancestor directory or mount point found in the catalog.
+  [:file => :path, :mount => :name].each do |type, property|
+    autorequire(type) do
+      req = []
+      path = Pathname.new(self[property])
+      if !path.root?
+        # Start at our parent, to avoid autorequiring ourself
+        parents = path.parent.enum_for(:ascend)
+        if found = parents.find { |p| catalog.resource(type, p.to_s) }
+          req << found.to_s
+        end
       end
+      # if the resource is a link, make sure the target is created first
+      req << self[:target] if self[:target] && type == :file
```

Comment:
Any explicit reason you don't require the mount happens before creating a link to the mount point?